### PR TITLE
Streamed signing

### PR DESCRIPTION
**Pull request by ni4**
This PR finishes implementation of the streamed approach, adding stremed signing, detached signing and cleartext signing.
Also it improves logic for output file creation, see the last commit/method rnp_initialize_io()

It doesn't remove old code - it will be removed as the separate PR.
Also some changes may happen later with CLI interface update - encrypt() and sign() methods will be merged to single encrypt_and_sign() once we'll have separate command line keys for recipients and signers.

### CODE REVIEW

**Comment by dewyatt** [comment]
This looks good overall, +1 on adding issuer fingerprint and moving keyid to unhashed. I had some minor comments, but don't really mind it being merged as-is either.

File: src/lib/misc.c [hunk 1]
```diff
@@ -1084,6 +1084,77 @@ rnp_file_exists(const char *path)
...
+    }
+
+    if (ext[0] == '.') {
+        ext++;
+    }
+
+    plen = strlen(path);
+    elen = strlen(ext);
+
+    return (elen < plen) && !strcmp(path + plen - elen, ext) && (path[plen - elen - 1] == '.');
```
[the remark this answers is not in the document]
Comment (ni4) [reply src/lib/misc.c]: Actually rnp_strcasecmp would be better, thanks. Despite the fact that `somefile.ASC` looks weird.

File: src/lib/rnp.c [hunk 2]
```diff
@@ -1438,6 +1468,46 @@ rnp_encrypt_stream(rnp_ctx_t *ctx, const char *in, const char *out)
...
+    handler->key_provider = &keyprov;
+    handler->ctx = ctx;
+    handler->param = NULL;
+
+    result = rnp_sign_src(handler, &src, &dst);
+    if (result != RNP_SUCCESS) {
+        RNP_LOG("failed with error code 0x%x", (int) result);
+    }
+
+finish:
```
[the remark this answers is not in the document]
Comment (ni4) [reply src/lib/rnp.c]: Agree, mislooked this. Fixed, thanks.

File: src/librepgp/stream-packet.c [hunk 3]
```diff
@@ -458,9 +483,136 @@ stream_write_one_pass(pgp_one_pass_sig_t *onepass, pgp_dest_t *dst)
...
+        return false;
+    }
+
+    /* add space for subpackets length */
+    res = add_packet_body_uint16(&spbody, 0);
+
+    for (list_item *sp = list_front(sig->subpkts); sp; sp = list_next(sp)) {
+        subpkt = (pgp_sig_subpkt_t *) sp;
+
+        if (subpkt->hashed != !!hashed) {
```
[the remark this answers is not in the document]
Comment (ni4) [reply src/librepgp/stream-packet.c]: Thanks, got over-defensive here and in few other places.

**Comment by ni4** [comment]
@dewyatt I updated PR with fixes according the comments. Are there any further comments, or this may be merged (despite Travis fails, #550 should be fixed at some point) ?

**Comment by dewyatt** [comment]
LGTM, feel free to merge.

**Comment by ni4** [comment]
Thanks, @dewyatt !

**Pull request merged**